### DISCUSSION TITLE
Simple Server polling external REST API

### DISCUSSION THREAD

**Discussion by Kinchkun** [tokio-rs/tokio#3531] [Q&A] [answered]
Hi,

I am totally new in async programming and need some starting point.

I would like to write a server which polls every 60 seconds (or so) an external REST API to fetch some data. Then process this data and again calls an external REST API with some commands.

While doing so I would like to be able to communicate with this server. For example I would like so signal that the server should shutdown. In order to shutdown I would like to cancel every outgoing REST request, update some file and exit the process.

I am making the rest calls using reqwest.

My very first approach was to:

```
loop {
   let data = update_data().await;
   let actions = process_data(data).await;
   execute(actions).await:
   wait().await; // use tokio::time::sleep
}
```

While this kinda works I cannot signal the server anything. I was trying to use tokio::signal::ctrl_c like so:

```
loop {
  register_ctrlc_handler();
   // same as above
}

async fn register_ctrlc_handler() {
   tokio::signal::ctrl_c()
            .await
            .expect("Could not register ctrc handler");
   info!("Terminating...")
  terminate();
}
```

But this approach does not work. I learned that async in Rust is lazy therefore the function `register_ctrlc_handler` is never executed

I am thinking that the Stream API could work, like:

```
while let Some(message) = stream.next().await {
        match message {
            Tick => execute_next_cycle(), // same thing as above, update_date, process_data...,
            Kill => terminate()
       }
}
```

Ideally this stream would react to Tcp-Events and signals (e.g. ctrlc). But I am clueless how to implement such a stream. 

I am thankful for every hint.

Kind regards
Kinch

**Comment by Darksonn** [comment, marked as answer]
To kill a currently running task, check out [this answer][1].

[1]: https://stackoverflow.com/questions/64084955/how-to-remotely-shut-down-running-tasks-with-tokio/64093362#64093362

**Reply by Kinchkun** [reply]
Thank you!
